Reduce memory allocations in HttpFileSystembasedFindPackageByIdResource

zivkan opened the pull request:
<!-- DO NOT MODIFY OR DELETE THIS TEMPLATE. IT IS USED IN AUTOMATION. -->

# Bug

<!-- If this is an engineering change or test change only, you do not need an issue. -->
<!-- Find or create an issue in NuGet/Home and paste the full url. -->
<!-- At the maintainers discretion, multiple changes may apply to a single issue, but only when the PRs are all created within a short period of time. -->
Fixes: https://github.com/NuGet/Home/issues/14095

## Description

Following @ToddGrun's PR: https://github.com/NuGet/NuGet.Client/pull/6265, in particular this comment thread: https://github.com/NuGet/NuGet.Client/pull/6265#discussion_r1980312868

I noticed that the code is creating one `PackageInfo` and `PackageIdentity` and `ContentUri` for every package version the server knows about. However, in a restore, typically only a single package version is needed. So, by creating those `PackageIdentity` and url strings only when needed, we can remove a whole lot of additional allocations over what the other PR did.

In addition to removing `PackageInfo` and associated types, I moved the JSON parsing to System.Text.Json instead of Newtonsoft.Json (which used JObject, no less).  So, that provides us with just most of Todd's PR memory improvements as well.

## PR Checklist

- [x] Meaningful title, helpful description and a linked NuGet/Home issue
- [x] ~Added tests~  code was refactored, no behavior changes
- [x] ~Link to an issue or pull request to update docs if this PR changes settings, environment variables, new feature, etc.~ N/A

zivkan committed:
Reduce memory allocations in HttpFileSystembasedFindPackageByIdResource

zivkan committed:
don't forget to sort

ToddGrun commented on src/NuGet.Core/NuGet.Protocol/RemoteRepositories/HttpFileSystemBasedFindPackageByIdResource.cs:
```diff
@@ -540,36 +545,27 @@ ex.InnerException is IOException &&
-            }
-
-            return streamResults;
+            builder.Sort();
```
>builder.Sort(); [](http://example.com/codeflow?start=12&length=15)

does it need to not add duplicates?

zivkan replied:
Our public documentation for the ["package content" resource](https://learn.microsoft.com/en-us/nuget/api/package-base-address-resource) states that to download a package, the URL is auto generated from the pattern: `{@id}/{LOWER_ID}/{LOWER_VERSION}/{LOWER_ID}.{LOWER_VERSION}.nupkg`.

If the version list contains duplicates, suggesting that there are two different packages with the same version, there's no way for NuGet to download the different versions. The download URLs will be identical. Hence servers can't have two different packages with the same version.

So, I consider it a server-side bug if there exists any NuGet server that returns duplicate versions in the index.json.

While the version list is returned to the caller in this class' `GetAllVersionsAsync` API, all the other APIs only check if the version exists. If there are two instances of the same version, the result will still be true, so I don't see any harm done. 

There's only a question if there is harm in `GetAllVersionsAsync` returning duplicates. That's certainly possible if someone calls `.ToDictionary` on the result, or uses `Dictionary.Add`, using the version as the key, or `.Single`, as these APIs will throw an exception.

But as long as there are no unhandled exceptions, maybe I'm not being customer focused enough, but I'm not worried about server version lists containing a duplicate.  I would feel comfortable telling the customer that the NuGet server they're using is providing incorrect results, which Visual Studio or the dotnet CLI is reporting accurately.

I don't know if anyone finds my justification convincing 😁

ToddGrun commented on src/NuGet.Core/NuGet.Protocol/RemoteRepositories/HttpFileSystemBasedFindPackageByIdResource.cs:
```diff
@@ -34,8 +37,8 @@ public class HttpFileSystemBasedFindPackageByIdResource : FindPackageByIdResourc
         private readonly HttpSource _httpSource;
-        private readonly ConcurrentDictionary<string, AsyncLazy<SortedDictionary<NuGetVersion, PackageInfo>>> _packageInfoCache =
-            new ConcurrentDictionary<string, AsyncLazy<SortedDictionary<NuGetVersion, PackageInfo>>>(StringComparer.OrdinalIgnoreCase);
+        private readonly ConcurrentDictionary<string, AsyncLazy<ImmutableArray<NuGetVersion>?>> _packageInfoCache =
```
>_packageInfoCache [](http://example.com/codeflow?start=96&length=17)

maybe _packageVersionCache now?

ToddGrun commented on src/NuGet.Core/NuGet.Protocol/RemoteRepositories/HttpFileSystemBasedFindPackageByIdResource.cs:
```diff
@@ -34,8 +37,8 @@ public class HttpFileSystemBasedFindPackageByIdResource : FindPackageByIdResourc
         private readonly HttpSource _httpSource;
-        private readonly ConcurrentDictionary<string, AsyncLazy<SortedDictionary<NuGetVersion, PackageInfo>>> _packageInfoCache =
-            new ConcurrentDictionary<string, AsyncLazy<SortedDictionary<NuGetVersion, PackageInfo>>>(StringComparer.OrdinalIgnoreCase);
+        private readonly ConcurrentDictionary<string, AsyncLazy<ImmutableArray<NuGetVersion>?>> _packageInfoCache =
```
>ImmutableArray<NuGetVersion>? [](http://example.com/codeflow?start=64&length=29)

Does this need to use a nullable ImmutableArray? Could the empty array just be used in those cases?

zivkan replied:
you're right, this can return an empty array

ToddGrun commented:
I like this PR, it was a good find that the PackageInfo didn't really need to be in this cache.

ToddGrun commented on src/NuGet.Core/NuGet.Protocol/RemoteRepositories/HttpFileSystemBasedFindPackageByIdResource.cs:
```diff
@@ -540,36 +545,27 @@ ex.InnerException is IOException &&
             string idInLowerCase = id.ToLowerInvariant();
+            var baseUri = _baseUris[retry % _baseUris.Count].OriginalString;
 
             var builder = StringBuilderPool.Shared.Rent(256);
```
>var builder = StringBuilderPool.Shared.Rent(256); [](http://example.com/codeflow?start=12&length=49)

When you debug this, do you see many of the url exceeding 256 and thus not working well with the pool?

zivkan replied:
While talking to Jeff, he pointed out that the string isn't dynamic in any way (no `if (something) builder.Add`), so we can use `string.Concat` instead.

ToddGrun replied:
It's still probably preferred over potentially removing items from the stringbuilder pool, but I think netfx will end up allocating the array on the heap to hold the individual strings.

zivkan committed:
Changes

GetAllVersionsAsync probably doesn't need to be sorted.
This lets us use a HashSet for quicker lookups in all other APIs.
Don't change the baseUri used to generate nupkg download URIs
String.Concat instead of StringBuilder

zivkan committed:
Update field name to be more accurate

ToddGrun commented on src/NuGet.Core/NuGet.Protocol/RemoteRepositories/HttpFileSystemBasedFindPackageByIdResource.cs:
```diff
@@ -480,13 +485,14 @@ private async Task<SortedDictionary<NuGetVersion, PackageInfo>> FindPackagesById
                         async httpSourceResult =>
                         {
-                            var result = new SortedDictionary<NuGetVersion, PackageInfo>();
+                            HashSet<NuGetVersion> result = [];
```
>HashSet<NuGetVersion> result = []; [](http://example.com/codeflow?start=28&length=34)

maybe leave this unitialized and set it to a new dictionary in an else clause below?

zivkan commented on src/NuGet.Core/NuGet.Protocol/RemoteRepositories/HttpFileSystemBasedFindPackageByIdResource.cs:
```diff
@@ -127,9 +130,9 @@ public override async Task<IEnumerable<NuGetVersion>> GetAllVersionsAsync(
+                var packageVersions = await GetAvailablePackageVersionsAsync(id, cacheContext, logger, cancellationToken);
 
-                return packageInfos.Keys;
+                return packageVersions;
```
Since `_packageVersionsCache` now contains `HashSet` rather than `SortedDictionary`, this means the public API `GetAllVersionsAsync` may no longer be returning a sorted list of versions.  So, changing this risks breaking behaviour.

However, `FindPackageById` is only used by restore, within NuGet.Client.  PM UI, and anything that cares about unlisted package status, does not use this API.  Restore handles this list not being sorted, and all tests appear to be passing, so I think the biggest risk are apps using our NuGet.Protocol package.

zivkan commented on src/NuGet.Core/NuGet.Protocol/RemoteRepositories/HttpFileSystemBasedFindPackageByIdResource.cs:
```diff
@@ -34,9 +36,10 @@ public class HttpFileSystemBasedFindPackageByIdResource : FindPackageByIdResourc
+        private readonly ConcurrentDictionary<string, AsyncLazy<HashSet<NuGetVersion>>> _packageVersionsCache =
+            new ConcurrentDictionary<string, AsyncLazy<HashSet<NuGetVersion>>>(StringComparer.OrdinalIgnoreCase);
         private readonly IReadOnlyList<Uri> _baseUris;
+        private string _chosenBaseUri;
```
I don't like this, I would have liked to make it round-robin load balance across all the base URIs that the server provided in their service index.  But that would be a change in behavior which might risk breaking customers.  So, we re-use the base uri that the version list was obtained from, just like the old code.

ToddGrun commented:
I'd love to hear about what measurements you ended up taking and how big of an improvement this is

ToddGrun commented on src/NuGet.Core/NuGet.Protocol/RemoteRepositories/HttpFileSystemBasedFindPackageByIdResource.cs:
```diff
@@ -537,68 +544,53 @@ ex.InnerException is IOException &&
-            };
+            var baseUri = _chosenBaseUri ?? _baseUris[0].OriginalString;
+
+            string contentUri = string.Concat(
```
>string contentUri = string.Concat( [](http://example.com/codeflow?start=12&length=34)

Would an interpolated string make this easier to read?

string contentUri = $"{baseUri}{idInLowerCase}/{normalizedVersionString}/{idInLowerCase}.{normalizedVersionString}.nupkg";

nkolev92 replied:
Isn't string.concat supposed to be faster than interpolation or format?

ToddGrun replied:
I was playing around on sharplab and they look pretty similar (at least on netfx) and although different on core, I'm not really sure that one is better than the other.

donnie-msft replied:
I agree that interpolation would be more readable, but this seems like a hotpath and a reasonably long string, so my understanding is Concat is more performant.

For the longest time, I understood concat or `+` for strings to be sloppy and bad performance. But a few years ago I learned that was somehow the opposite of reality!

Therefore, if my understanding is still up to date, I'd say leave it as is or compromise for readability with:
`baseUri +  idInLowerCase + "/" + normalizedVersionString + "/" + idInLowerCase ...`

ToddGrun replied:
I'm fine with whatever is decided here, but I do think the perf of interpolated strings is better than you guys think:

A [simple benchmark](https://gist.github.com/ToddGrun/7320d4d7f535496ba540af2ca2ab6c96) comparing String.Concat and interpolation actually indicates interpolation performance is about the same on net472 and actually better than String.Concat in net9.

net472
|                         Method |     Mean |   Error |  StdDev |   Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------------------- |---------:|--------:|--------:|--------:|------:|------:|----------:|
|                         Concat | 150.5 us | 1.56 us | 1.46 us | 54.9316 |     - |     - |    282 KB |
|       ConcatIncludeLoopCounter | 233.9 us | 3.23 us | 3.03 us | 63.2324 |     - |     - |    325 KB |
|                   Interpolated | 147.3 us | 0.87 us | 0.81 us | 54.9316 |     - |     - |    282 KB |
| InterpolatedIncludeLoopCounter | 162.0 us | 1.14 us | 1.01 us | 56.3965 |     - |     - |    289 KB |

net9.0
|                         Method |     Mean |    Error |   StdDev |   Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------------------- |---------:|---------:|---------:|--------:|------:|------:|----------:|
|                         Concat | 43.86 us | 0.642 us | 0.536 us | 28.2593 |     - |     - |    289 KB |
|       ConcatIncludeLoopCounter | 73.05 us | 1.286 us | 1.140 us | 44.1895 |     - |     - |    451 KB |
|                   Interpolated | 38.67 us | 0.354 us | 0.331 us | 19.1040 |     - |     - |    195 KB |
| InterpolatedIncludeLoopCounter | 42.38 us | 0.433 us | 0.384 us | 19.1040 |     - |     - |    195 KB |

zivkan commented:
> I'd love to hear about what measurements you ended up taking and how big of an improvement this is

From: https://github.com/NuGet/NuGet.Client/pull/6265#discussion_r1980768945

|branch|GC stats total allocs|GC Heap Alloc Ignore Free Stacks ConsumeFlatContainer|
|--|--|--|
|dev|252 MB|5,512,529|
|this PR|246 MB|2,139,213|
|my PR|244 MB|669,655|

But keep in mind in the comment, from the other PR, I had the caveat that I amit I pretty much don't know what I'm doing.

Here's a PerfView diff of the dev branch and this PR branch heap alloc view:
![image](https://github.com/user-attachments/assets/a3b6971b-4c0e-4d71-ba86-df4d4fde6336)

Finally, I thought I'd try to use BenchmarkDotNet's MemoryAnalyzer. Here's a patch to the NuGet.Client repo with my benchmark code: 
[benchmark.zip](https://github.com/user-attachments/files/19116333/benchmark.zip)

However, when I run it, I get these results:

| Method   | Mean     | Error     | StdDev    | Ratio | Gen0     | Gen1    | Allocated | Alloc Ratio |
|--------- |---------:|----------:|----------:|------:|---------:|--------:|----------:|------------:|
| Change   | 2.739 ms | 0.0161 ms | 0.0126 ms |  1.00 | 328.1250 | 23.4375 |   1.99 MB |        1.00 |
| Original | 2.656 ms | 0.0246 ms | 0.0205 ms |  0.97 | 316.4063 | 23.4375 |   1.91 MB |        0.96 |

So, this benchmark results make it look like my change has more allocations and is slower than the original, which frankly doesn't make any sense to me whatsoever.

What started as reviewing another PR turned into trying to implement further improvements, and while code changes took probably less than an hour, I've now been working probably 2 full days on collecting perf measurements.  I was behind my assigned work for this sprint before I started any of this, so I just can't justify any more effort.  If this PR is "acceptable" as-is, or with minor changes, I can do that.  But I don't have the capacity to investigate any further why these perf measurements are so out of whack.

ToddGrun commented:
I'm very happy with this PR (for what that's worth). I definitely wasn't trying to steal a bunch of anyone on the nuget's team time (which is why I made the original PR myself). I went ahead and did some measurements based on this branch, and will share below. My measurements were based on the following steps: "git clean -xdff" in roslyn repo, delete the global nuget cache, open VS, start profiler, and open roslyn.sln, wait for the status bar to indicate everything is done, stop profiler.

I took two profiles each with and without this change and averaged the allocations/times. Filtering under FindPackagesByIdAsync, this change showed an allocation improve from 155 MB to 25 MB and CPU samples improve from 21000 ms to 3850 ms.

---
In reply to: [2705152207](https://github.com/NuGet/NuGet.Client/pull/6300#issuecomment-2705152207) [](http://example.com/codeflow?ancestors=2705152207)

ToddGrun commented on src/NuGet.Core/NuGet.Protocol/RemoteRepositories/HttpFileSystemBasedFindPackageByIdResource.cs:
```diff
@@ -537,68 +544,53 @@ ex.InnerException is IOException &&
+        private static async Task<HashSet<NuGetVersion>> ConsumeFlatContainerIndexAsync(Stream stream, string id, string baseUri, CancellationToken token)
         {
-            var doc = await stream.AsJObjectAsync(token);
+            var json = await JsonSerializer.DeserializeAsync<FlatContainerVersionList>(stream, cancellationToken: token);
```
>JsonSerializer.DeserializeAsync [](http://example.com/codeflow?start=29&length=31)

This shows up as a very small amount of the remaining allocations, so there is definitely no need to pursue the other PR.

donnie-msft replied:
👏

zivkan commented:
While falling asleep last night, I realized that my benchmark was reusing a single `SourceRepository` instance, which means the same `HttpFileSystemBasedFindPackageByIdResource` instance, which means the warmup would warm the cache, and then the benchmark was only testing the cache-hit code path!  It also explains why my change was slower and using more memory, because the old code pre-cached the content URL and PackageIdentity for every package version, whereas my change allocates those types on demand.

Fixing the benchmark, I get different results:

| Method   | Mean     | Error   | StdDev  | Ratio | RatioSD | Gen0      | Gen1     | Allocated | Alloc Ratio |
|--------- |---------:|--------:|--------:|------:|--------:|----------:|---------:|----------:|------------:|
| Change   | 206.4 ms | 4.11 ms | 5.20 ms |  1.00 |    0.03 | 1000.0000 |        - |  20.67 MB |        1.00 |
| Original | 201.5 ms | 4.02 ms | 7.25 ms |  0.98 |    0.04 | 1333.3333 | 666.6667 |  24.77 MB |        1.20 |

I'm still very surprised the changed code is slower, wall clock time. But at least we're now correctly measuring the memory allocation improvements.

zivkan committed:
minor improvements

donnie-msft commented on src/NuGet.Core/NuGet.Protocol/RemoteRepositories/HttpFileSystemBasedFindPackageByIdResource.cs:
```diff
@@ -540,65 +547,50 @@ ex.InnerException is IOException &&
+        private static async Task<HashSet<NuGetVersion>> ConsumeFlatContainerIndexAsync(Stream stream, string id, string baseUri, CancellationToken token)
         {
-            var doc = await stream.AsJObjectAsync(token);
+            var json = await JsonSerializer.DeserializeAsync<FlatContainerVersionList>(stream, cancellationToken: token);
```
nit: just name it `cancellationToken`?

zivkan replied:
if you're suggesting to remove the named paramter, that's not possible because the method signature is [`DeserializeAsync<TValue>(System.IO.Stream utf8Json, System.Text.Json.JsonSerializerOptions? options = default, System.Threading.CancellationToken cancellationToken = default)`](https://learn.microsoft.com/en-us/dotnet/api/system.text.json.jsonserializer.deserializeasync?view=net-9.0#system-text-json-jsonserializer-deserializeasync-1(system-io-stream-system-text-json-jsonserializeroptions-system-threading-cancellationtoken))

Since I'm not providing a value for the second parameter (`options`), I need to tell the compiler I intend on passing the token to the `cancellationToken` paramter, so the compiler doesn't have a type mismatch error.

donnie-msft replied:
Yes, I didn't catch that it was an optional argument, so that's why I suggested it. Makes sense.
It still seems worth renaming this to match other places that use the same name, but now the motivation is lower.

donnie-msft commented on src/NuGet.Core/NuGet.Protocol/RemoteRepositories/HttpFileSystemBasedFindPackageByIdResource.cs:
```diff
@@ -192,22 +195,24 @@ public override async Task<FindPackageByIdDependencyInfo> GetDependencyInfoAsync
-                        cancellationToken);
-
-                    return GetDependencyInfo(reader);
+                    // package version not found
```
nit: consider omitting comment - the code seems clear

donnie-msft commented on src/NuGet.Core/NuGet.Protocol/RemoteRepositories/HttpFileSystemBasedFindPackageByIdResource.cs:
```diff
@@ -278,21 +283,23 @@ public override async Task<bool> CopyNupkgToStreamAsync(
                 cancellationToken.ThrowIfCancellationRequested();
 
-                var packageInfos = await EnsurePackagesAsync(id, cacheContext, logger, cancellationToken);
+                var packageVersions = await GetAvailablePackageVersionsAsync(id, cacheContext, logger, cancellationToken);
```
nit: Please use an explicit type so readers immediately know this is a hashset.

donnie-msft commented on src/NuGet.Core/NuGet.Protocol/RemoteRepositories/HttpFileSystemBasedFindPackageByIdResource.cs:
```diff
@@ -540,65 +547,50 @@ ex.InnerException is IOException &&
         {
-            var parsedVersion = NuGetVersion.Parse(version);
-            var normalizedVersionString = parsedVersion.ToNormalizedString();
+            var normalizedVersionString = version.ToNormalizedString().ToLowerInvariant();
```
This seems reasonable given we're calling ToLower on the Id as well.
However, can the tests be updated to match this? If I'm not mistaken, this test would fail after your change if the test data were uppercase (but it's not, so it won't fail now). Just making 1 uppercase character and amending this assertion would cover that?
https://github.com/NuGet/NuGet.Client/blob/d948b1a55a189466ccb627c7f754226d9f43505f/test/NuGet.Core.Tests/NuGet.Protocol.Tests/RemoteRepositories/HttpFileSystemBasedFindPackageByIdResourceTests.cs#L328C27-L328C28

zivkan replied:
There was already a test failing, which is why I needed to add this in the first place: https://github.com/NuGet/NuGet.Client/blob/13aaaf31db2958ca85b0b98b8dea8fcf6a7961de/test/NuGet.Core.Tests/NuGet.Protocol.Tests/RemoteRepositories/HttpFileSystemBasedFindPackageByIdResourceTests.cs#L318-L323

It actually took me a long time to figure out why the test was passing previously. Turns out that since the old code was pre-caching nupkg URLs for all package versions, it was using the version string as provided by the server's index.json version list.

Writing this out reminded me there's a theoretical regression. I updated the PR, but couldn't maintain the legacy behaviour with netstandard2.0. I added a comment explaining it.

donnie-msft replied:
I didn't notice there was an uppercase in this test already. 
Thanks for adding the comment.

donnie-msft reviewed (commented):
Nice work, and I look forward to seeing how perf improves as a result of this PR!

Not sure if I feel reassured or worried about test coverage given that no tests needed to be changed here. I did suggest 1 string related test change.

zivkan committed:
More improvements

zivkan committed:
Add null annotations

zivkan committed:
fix bug

ToddGrun commented on src/NuGet.Core/NuGet.Protocol/RemoteRepositories/HttpFileSystemBasedFindPackageByIdResource.cs:
```diff
@@ -418,16 +425,45 @@ public override async Task<bool> DoesPackageExistAsync(
+                versionString = null;
+                return false;
+            }
+            versionString = version.ToNormalizedString().ToLowerInvariant();
```
>version.ToNormalizedString().ToLowerInvariant(); [](http://example.com/codeflow?start=28&length=48)

If the !NETSTANDARD case can get away without needing the ToLowerInvariant, couldn't the else clause versionString assignment be changed to:

versionString = packageVersions[version].ToNormalizedString();

zivkan replied:
Unfortunately, we can't remove the `ToLowerInvariant()`, for the reason described here: https://github.com/NuGet/NuGet.Client/pull/6300#discussion_r1986961148.  Perhaps I need to improve the comment at the top of the NETSTANDARD block

ToddGrun replied:
I'm pretty sure I'm being super dense. I definitely believe that the ToLowerInvariant can be needed, I just don't understand why the !NETSTANDARD block wouldn't need it then.

ToddGrun replied:
Just to make sure we're on the same page, I was suggesting that the NETSTANDARD case also use the value from the HashSet, not from the input version.

zivkan replied:
Unless there's [an API I'm not aware of](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.hashset-1?view=net-9.0#methods), the only way to get the value out of the hashset on netstandard is to iterate the hashset as if it was a list and stop when the value is equal.  `HashSet<T>` on netstandard doesn't have a `TryGetValue` API, like the comment above the code says.

But yes, if there is an easy/cheap way to get the value out of the hashset, then I'd have it work the same way as netframework and netcoreapp.

ToddGrun replied:
Gotcha, it was very stuck in my mind that there was an indexer on HashSet. Thanks for the clarification!

ToddGrun reviewed (approved):
:shipit:

zivkan committed:
HttpSource.GetAsync needs to ensure that files are closed before returning

zivkan commented:
@NuGet/nuget-client tests have been fixed. This is ready to merge now.

donnie-msft commented on src/NuGet.Core/NuGet.Protocol/HttpSource/HttpSource.cs:
```diff
@@ -81,116 +81,125 @@ public virtual async Task<T> GetAsync<T>(
+                            }
+                            catch (Exception e)
+                            {
+                                cacheResult.Stream.Dispose();
```
I know we see a lot of ObjectDisposedExceptions in telemetry, but I'm not sure how impactful those are. Could this be a place where checking the stream is not null & not disposed would ensure customers see the error message in the log?

zivkan replied:
How do you check if a stream has already been disposed?

The [`System.IO.Stream` abstract class](https://learn.microsoft.com/en-us/dotnet/api/system.io.stream?view=net-9.0) doesn't have any APIs I can see. Neither does [`FileStream`](https://learn.microsoft.com/en-us/dotnet/api/system.io.filestream?view=net-9.0). For callers of NuGet.Protocol that use `SourceCacheContext.DirectDownload = true`, the stream ends up being an internal class defined in System.Net.Http.dll, so we wouldn't be able to cast to that type to call APIs on it either, without reflection.

Finally, if you test FileStream, and HttpClient with both values of `HttpCompletionOption`, and call dispose multiple times on `HttpResponseMessage.Content.ReadAsStreamAsync()`, no exception is thrown.

zivkan replied:
I'm resolving this, to merge the PR. Reach out to me if there are followups, and I can update in another PR.

donnie-msft commented on src/NuGet.Core/NuGet.Protocol/RemoteRepositories/HttpFileSystemBasedFindPackageByIdResource.cs:
```diff
@@ -418,16 +425,45 @@ public override async Task<bool> DoesPackageExistAsync(
+#else
+            // .NET Standard 2.0 doesn't support HashSet<T>.TryGetValue. The API docs say the version should be lowercase, but if
+            // a server doesn't follow the spec and uses case sensitive URLs, this might result in 404 responses.
+            if (!packageVersions.Contains(version))
```
Thanks for adding the comment. This makes sense now: We will find the version regardless of casing, then make it lowercase, then make the request which could be a 404. At least there's a chance someone using a case-sensitive server would notice the mismatch in casing in the URL.
